src/core/license.rs

```diff
@@ -63,82 +96,914 @@ pub enum StateTransition {
+            return Err(Error::RdpError(RdpError::new(
+                RdpErrorKind::InvalidData,
+                "SEC: Invalid Licence packet",
+            )));
+        }
+
+        let mut license_message = component![
+            "bMsgtype" => 0_u8,
+            "flag" => Check::new(Preambule::PreambleVersion30 as u8),
+            "wMsgSize" => DynOption::new(U16::LE(0), |size| MessageOption::Size("message".to_string(), size.inner() as usize - 4)),
+            "message" => Vec::<u8>::new()
+        ];
+
+        license_message.read(&mut stream)?;
+        let msg_type = cast!(DataType::U8, license_message["bMsgtype"])?;
+        let mut msg_data = Cursor::new(cast!(DataType::Slice, license_message["message"])?);
+
+        match MessageType::try_from(cast!(DataType::U8, license_message["bMsgtype"])?)? {
+            MessageType::NewLicense => Ok(Self::NewLicense(NewLicense::from_bytes(&mut msg_data)?)),
+            MessageType::LicenseRequest => Ok(Self::LicenseRequest(
+                ServerLicenseRequest::from_bytes(&mut msg_data)?,
+            )),
+            MessageType::PlatformChallenge => Ok(Self::PlatformChallenge(
+                PlatformChallenge::from_bytes(&mut msg_data)?,
+            )),
+            MessageType::UpgradeLicense => Ok(Self::UpgradeLicense(UpgradeLicense::from_bytes(
+                &mut msg_data,
+            )?)),
+            MessageType::ErrorAlert => Ok(Self::ErrorAlert(ErrorAlert::from_bytes(&mut msg_data)?)),
+            _ => Err(Error::RdpError(RdpError::new(
+                RdpErrorKind::NotImplemented,
+                &format!("Licensing nego not implemented. bMsgtype: {:?}", msg_type),
+            ))),
+        }
+    }
+}
+
+pub enum ServerCertificate {
+    Proprietary(RsaPublicKey),
+    X509(RsaPublicKey),
+}
+
+impl ServerCertificate {
+    fn from_bytes(raw: &mut dyn Read) -> RdpResult<Self> {
+        let mut binary_blob_server_certificate = component![
+            "dwVersion" => U32::LE(0),
+            "certData" => Vec::<u8>::new()
+        ];
+        binary_blob_server_certificate.read(raw)?;
+        let cert_version = cast!(DataType::U32, binary_blob_server_certificate["dwVersion"])?;
+        let mut cert_data = cast!(DataType::Slice, binary_blob_server_certificate["certData"])?;
+
+        match cert_version & CERT_CHAIN_VERSION_MASK {
+            CERT_CHAIN_VERSION_1 => {
+                let mut message = component![
+                    "dwSigAlgId" => U32::LE(0),
+                    "dwKeyAlgId" => U32::LE(0),
+                    "wPublicKeyBlobType" => U16::LE(0),
+                    "wPublicKeyBlobLen" => DynOption::new(U16::LE(0), | size | MessageOption::Size("PublicKeyBlob".to_string(), size.inner() as usize)),
+                    "PublicKeyBlob" => component![
+                        "magic" => U32::LE(0),
+                        "keylen" => DynOption::new(U32::LE(0), | size | MessageOption::Size("modulus".to_string(), size.inner() as usize - 8)),
```

Comment:
What's the `-8` for here?

Reply:
I should've make it const and give it a name, nice catch!

It's just a value that is used to calculate `keylen`. There is a padding which we must take into account when we calculate the value of this field  so that's why we need to subtract the `8`.

Here's the spec file for that data structure:
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/fe93545c-772a-4ade-9d02-ad1e0d81b6af